mirrord/console/src/main.rs

```diff
@@ -0,0 +1,111 @@
+mod protocol;
+
+use std::any::type_name;
+
+use axum::{
+    extract::ws::{Message, WebSocket, WebSocketUpgrade},
+    response::Response,
+    routing::get,
+    Router,
+};
+use protocol::Hello;
+use tracing::{error, info, metadata::LevelFilter};
+use tracing_subscriber::{fmt::format::FmtSpan, prelude::*};
+
+struct WsWrapper {
+    socket: WebSocket,
+}
+
+impl WsWrapper {
+    fn new(socket: WebSocket) -> Self {
+        Self { socket }
+    }
+
+    async fn next_message<T>(&mut self) -> Option<T>
+    where
+        T: serde::de::DeserializeOwned,
+    {
+        let raw_msg = self.socket.recv().await;
+        let msg = match raw_msg {
+            None => {
+                error!("Client disconnected");
+                return None;
+            }
+            Some(Ok(Message::Binary(msg))) => msg,
+            _ => {
+                error!("Invalid message {raw_msg:?}");
+                return None;
+            }
+        };
+
+        match serde_json::from_slice::<T>(&msg) {
+            Ok(record) => Some(record),
+            Err(e) => {
+                let type_name = type_name::<T>();
+                error!("Client message error: {e:?}, was expecting {type_name:?}");
+                None
+            }
+        }
+    }
+}
+
+async fn handle_socket(socket: WebSocket) {
+    let mut wrapper = WsWrapper::new(socket);
+
+    let client_info: Hello = match wrapper.next_message().await {
+        Some(hello) => {
+            info!("Client connected -  process info {hello:?}");
+            hello
+        }
+        None => {
+            error!("Client disconnected");
+            return;
```

Comment:
Is there a reconnect mechanism? If there isn't, don't we want to crash here?

Reply:
it's logging, I don't think a crash is necessary.